File: apps/desktop/src/renderer/services/standingTasks.ts
/**
 * Standing tasks / scheduled automations REST client (L1 + L2a Webhook).
 *
 * Wire shapes come from OpenAPI (`@agentcore/contract-rest-types` via
 * `@/types/api.generated`). Domain models stay camelCase like `folders` /
 * `handoff`.
 */

import { BASE_URL, api } from "@/services/api";
import type { PermissionAxes } from "@/services/permissionAxes";
import {
  DEFAULT_PERMISSION_AXES,
  normalizeAxes,
} from "@/services/permissionAxes";
import type { components } from "@/types/api.generated";

type Schemas = components["schemas"];

type StandingTaskWire = Schemas["StandingTaskSummary"];
type StandingTaskRunWire = Schemas["StandingTaskRunSummary"];
type StandingTaskRunListWire = Schemas["StandingTaskRunListResponse"];
type CreateStandingTaskWire = Schemas["CreateStandingTaskRequest"];
type UpdateStandingTaskWire = Schemas["UpdateStandingTaskRequest"];
type TriggerStandingTaskWire = Schemas["TriggerStandingTaskResponse"];
type RotateWebhookSecretWire = Schemas["RotateWebhookSecretResponse"];

/** Built-in schedule presets (UI + create/patch). Custom uses `cron`. */
export type SchedulePreset =
  | "daily"
  | "weekdays"
  | "weekly_mon"
  | "weekly_fri"
  | "monthly_1"
  | "custom";

/** Per-task trigger; mutually exclusive (定案 L2a). */
export type TriggerKind = StandingTaskWire["trigger_kind"];

/** Optional run provenance for inbox display. */
export type TriggerSource = StandingTaskRunWire["trigger_source"];

export const SCHEDULE_PRESET_ORDER: SchedulePreset[] = [
  "daily",
  "weekdays",
  "weekly_mon",
  "weekly_fri",
  "monthly_1",
  "custom",
];

export const SCHEDULE_PRESET_LABELS: Record<SchedulePreset, string> = {
  daily: "每天",
  weekdays: "工作日",
  weekly_mon: "每周一",
  weekly_fri: "每周五",
  monthly_1: "每月 1 日",
  custom: "自定义 cron",
};

export const TRIGGER_KIND_ORDER: TriggerKind[] = ["schedule", "webhook"];

export const TRIGGER_KIND_LABELS: Record<TriggerKind, string> = {
  schedule: "定时",
  webhook: "Webhook",
};

export const TRIGGER_SOURCE_LABELS: Record<TriggerSource, string> = {
  schedule: "定时",
  webhook: "Webhook",
  manual: "手动",
};

export type StandingTaskRunStatus = StandingTaskRunWire["status"];

export interface StandingTask {
  id: string;
  name: string;
  triggerKind: TriggerKind;
  schedulePreset: SchedulePreset | null;
  cron: string | null;
  folderId: string;
  goal: string;
  permissionAxes: PermissionAxes;
  enabled: boolean;
  nextRunAt: string | null;
  conversationId: string | null;
  lastRunAt: string | null;
  webhookId: string | null;
  /** Public POST URL when trigger is webhook; may be absolute or path. */
  webhookUrl: string | null;
  /**
   * One-time plaintext secret — only present on create / rotate responses.
   * List/GET never return this; treat as ephemeral UI state.
   */
  webhookSecret: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface StandingTaskRun {
  id: string;
  standingTaskId: string;
  /** Denormalized task name when the API includes it; else null. */
  taskName: string | null;
  status: StandingTaskRunStatus;
  conversationId: string | null;
  userMessageId: string | null;
  summary: string | null;
  error: string | null;
  /** Null = unread / not dismissed. */
  ackedAt: string | null;
  /** Present when backend sends it; inbox may show a source chip. */
  triggerSource: TriggerSource | null;
  createdAt: string;
  finishedAt: string | null;
}

export interface CreateStandingTaskInput {
  name: string;
  triggerKind: TriggerKind;
  /** Required when `triggerKind === "schedule"`. */
  schedulePreset?: SchedulePreset;
  cron?: string | null;
  folderId: string;
  goal: string;
  permissionAxes: PermissionAxes;
  enabled?: boolean;
}

export interface PatchStandingTaskInput {
  name?: string;
  triggerKind?: TriggerKind;
  schedulePreset?: SchedulePreset | null;
  cron?: string | null;
  folderId?: string;
  goal?: string;
  permissionAxes?: PermissionAxes;
  enabled?: boolean;
}

export interface ListStandingTaskRunsQuery {
  status?: StandingTaskRunStatus | StandingTaskRunStatus[];
  /** Prefer unread badge rows when the API supports it. */
  unacked?: boolean;
  limit?: number;
}

export interface RotateWebhookSecretResult {
  webhookSecret: string;
  webhookUrl: string | null;
  webhookId: string | null;
}

/** Immediate-run response — OpenAPI `TriggerStandingTaskResponse`. */
export interface TriggerStandingTaskResult {
  runId: string;
}

function asPreset(raw: string | null | undefined): SchedulePreset | null {
  if (!raw) return null;
  return (SCHEDULE_PRESET_ORDER as string[]).includes(raw)
    ? (raw as SchedulePreset)
    : "custom";
}

function asTriggerKind(raw: string | null | undefined): TriggerKind {
  return raw === "webhook" ? "webhook" : "schedule";
}

function asTriggerSource(
  raw: string | null | undefined,
): TriggerSource | null {
  if (raw === "schedule" || raw === "webhook" || raw === "manual") return raw;
  return null;
}

/** Backend returns a path; external callers need an absolute URL. */
export function absoluteWebhookUrl(
  raw: string | null | undefined,
): string | null {
  if (!raw) return null;
  if (/^https?:\/\//i.test(raw)) return raw;
  const base = BASE_URL.replace(/\/$/, "");
  const path = raw.startsWith("/") ? raw : `/${raw}`;
  return `${base}${path}`;
}

function asRunStatus(raw: string): StandingTaskRunStatus {
  switch (raw) {
    case "running":
    case "succeeded":
    case "failed":
    case "awaiting_user":
      return raw;
    default:
      return "failed";
  }
}

export function toStandingTask(w: StandingTaskWire): StandingTask {
  return {
    id: w.id,
    name: w.name,
    triggerKind: asTriggerKind(w.trigger_kind),
    schedulePreset: asPreset(w.schedule_preset),
    cron: w.cron ?? null,
    folderId: w.folder_id,
    goal: w.goal,
    permissionAxes: normalizeAxes(w.permission_axes ?? DEFAULT_PERMISSION_AXES),
    enabled: w.enabled,
    nextRunAt: w.next_run_at ?? null,
    conversationId: w.conversation_id ?? null,
    lastRunAt: w.last_run_at ?? null,
    webhookId: w.webhook_id ?? null,
    webhookUrl: absoluteWebhookUrl(w.webhook_url),
    webhookSecret: w.webhook_secret ?? null,
    createdAt: w.created_at,
    updatedAt: w.updated_at,
  };
}

export function toStandingTaskRun(w: StandingTaskRunWire): StandingTaskRun {
  return {
    id: w.id,
    standingTaskId: w.standing_task_id,
    taskName: w.task_name ?? null,
    status: asRunStatus(w.status),
    conversationId: w.conversation_id ?? null,
    userMessageId: w.user_message_id ?? null,
    summary: w.summary ?? null,
    error: w.error ?? null,
    ackedAt: w.acked_at ?? null,
    triggerSource: asTriggerSource(w.trigger_source),
    createdAt: w.created_at,
    finishedAt: w.finished_at ?? null,
  };
}

function createBody(input: CreateStandingTaskInput): CreateStandingTaskWire {
  const kind = input.triggerKind;
  const body: CreateStandingTaskWire = {
    name: input.name,
    trigger_kind: kind,
    folder_id: input.folderId,
    goal: input.goal,
    permission_axes: input.permissionAxes,
    enabled: input.enabled ?? true,
  };
  if (kind === "schedule") {
    const preset = input.schedulePreset ?? "weekly_mon";
    body.schedule_preset = preset;
    // Named presets must not send cron (backend rejects both); custom requires it.
    if (preset === "custom") {
      body.cron = input.cron ?? null;
    }
  }
  return body;
}

function patchBody(input: PatchStandingTaskInput): UpdateStandingTaskWire {
  const body: UpdateStandingTaskWire = {};
  if (input.name !== undefined) body.name = input.name;
  if (input.triggerKind !== undefined) body.trigger_kind = input.triggerKind;

  const kind = input.triggerKind;
  if (kind === "webhook") {
    // Switching to / staying on webhook: do not send schedule fields.
  } else if (kind === "schedule" || input.schedulePreset !== undefined) {
    if (input.schedulePreset !== undefined && input.schedulePreset !== null) {
      body.schedule_preset = input.schedulePreset;
      if (input.schedulePreset === "custom" && input.cron !== undefined) {
        body.cron = input.cron;
      }
    } else if (input.cron !== undefined && input.schedulePreset === undefined) {
      body.cron = input.cron;
    }
  } else if (input.cron !== undefined) {
    body.cron = input.cron;
  }

  if (input.folderId !== undefined) body.folder_id = input.folderId;
  if (input.goal !== undefined) body.goal = input.goal;
  if (input.permissionAxes !== undefined)
    body.permission_axes = input.permissionAxes;
  if (input.enabled !== undefined) body.enabled = input.enabled;
  return body;
}

/** List the signed-in user's standing tasks. */
export async function listStandingTasks(): Promise<StandingTask[]> {
  const res = await api.get<StandingTaskWire[]>("/v1/standing-tasks");
  return (Array.isArray(res) ? res : []).map(toStandingTask);
}

export async function getStandingTask(id: string): Promise<StandingTask> {
  const res = await api.get<StandingTaskWire>(`/v1/standing-tasks/${id}`);
  return toStandingTask(res);
}

export async function createStandingTask(
  input: CreateStandingTaskInput,
): Promise<StandingTask> {
  const res = await api.post<StandingTaskWire>(
    "/v1/standing-tasks",
    createBody(input),
  );
  return toStandingTask(res);
}

export async function patchStandingTask(
  id: string,
  input: PatchStandingTaskInput,
): Promise<StandingTask> {
  const res = await api.patch<StandingTaskWire>(
    `/v1/standing-tasks/${id}`,
    patchBody(input),
  );
  return toStandingTask(res);
}

export async function deleteStandingTask(id: string): Promise<void> {
  await api.delete(`/v1/standing-tasks/${id}`);
}

/**
 * Trigger one run now. Response is OpenAPI `TriggerStandingTaskResponse`
 * (`run_id` only) — not a full inbox row.
 */
export async function runStandingTaskNow(
  id: string,
): Promise<TriggerStandingTaskResult> {
  const res = await api.post<TriggerStandingTaskWire>(
    `/v1/standing-tasks/${id}/run`,
    {},
  );
  if (!res?.run_id) {
    throw new Error("trigger standing task response missing run_id");
  }
  return { runId: res.run_id };
}

/**
 * Rotate webhook secret. Plaintext `webhook_secret` is returned once
 * (`RotateWebhookSecretResponse`).
 */
export async function rotateWebhookSecret(
  id: string,
): Promise<RotateWebhookSecretResult> {
  const res = await api.post<RotateWebhookSecretWire>(
    `/v1/standing-tasks/${id}/rotate-webhook-secret`,
    {},
  );
  if (!res?.webhook_secret) {
    throw new Error("rotate-webhook-secret response missing webhook_secret");
  }
  return {
    webhookSecret: res.webhook_secret,
    webhookUrl: absoluteWebhookUrl(res.webhook_url),
    webhookId: res.webhook_id,
  };
}

function runsQueryString(q: ListStandingTaskRunsQuery = {}): string {
  const params = new URLSearchParams();
  if (q.status !== undefined) {
    const statuses = Array.isArray(q.status) ? q.status : [q.status];
    for (const s of statuses) params.append("status", s);
  }
  if (q.unacked === true) params.set("unacked", "true");
  if (q.limit !== undefined) params.set("limit", String(q.limit));
  const qs = params.toString();
  return qs ? `?${qs}` : "";
}

export async function listStandingTaskRuns(
  query: ListStandingTaskRunsQuery = {},
): Promise<StandingTaskRun[]> {
  const res = await api.get<StandingTaskRunListWire>(
    `/v1/standing-task-runs${runsQueryString(query)}`,
  );
  return (res?.items ?? []).map(toStandingTaskRun);
}

/** Mark a run card read / dismiss a failure card. */
export async function ackStandingTaskRun(id: string): Promise<StandingTaskRun> {
  const res = await api.post<StandingTaskRunWire>(
    `/v1/standing-task-runs/${id}/ack`,
    {},
  );
  return toStandingTaskRun(res);
}

/**
 * Badge = unacked awaiting_user + unacked failed.
 * Prefer server ``badge`` on the list payload.
 */
export async function countInboxBadge(): Promise<number> {
  const res = await api.get<StandingTaskRunListWire>(
    "/v1/standing-task-runs?limit=1",
  );
  if (typeof res?.badge === "number" && Number.isFinite(res.badge)) {
    return res.badge;
  }
  return (res?.items ?? [])
    .map(toStandingTaskRun)
    .filter((r) => {
      if (r.status === "awaiting_user" && !r.ackedAt) return true;
      if (r.status === "failed" && !r.ackedAt) return true;
      return false;
    }).length;
}

/** List / editor subtitle for trigger. */
export function scheduleLabel(task: StandingTask): string {
  if (task.triggerKind === "webhook") return TRIGGER_KIND_LABELS.webhook;
  if (task.schedulePreset && task.schedulePreset !== "custom") {
    return SCHEDULE_PRESET_LABELS[task.schedulePreset];
  }
  if (task.cron) return task.cron;
  if (task.schedulePreset === "custom") return "自定义";
  return "未设置周期";
}

export function triggerSourceLabel(source: TriggerSource): string {
  return TRIGGER_SOURCE_LABELS[source];
}

File: apps/server/agentcore/api/routes/standing_tasks.py
"""Standing tasks + inbox + L2a webhook hook routes."""

from datetime import UTC, datetime
from typing import Literal

from fastapi import APIRouter, Depends, Header, Query, Request

from agentcore.api.dependencies import (
    AuthUser,
    get_folder_repo,
    get_standing_task_repo,
    get_standing_task_run_repo,
)
from agentcore.api.schemas import StatusResponse
from agentcore.api.schemas.standing_tasks import (
    CreateStandingTaskRequest,
    RotateWebhookSecretResponse,
    StandingTaskRunListResponse,
    StandingTaskRunSummary,
    StandingTaskSummary,
    TriggerStandingTaskResponse,
    UpdateStandingTaskRequest,
)
from agentcore.core.errors import NotFoundError, ValidationError
from agentcore.core.types import DEFAULT_PERMISSION_AXES, new_id
from agentcore.db.repositories import FolderRepository
from agentcore.db.repositories.standing_tasks import (
    StandingTaskRepository,
    StandingTaskRunRepository,
)
from agentcore.standing_tasks.paths import webhook_path
from agentcore.standing_tasks.runner import dispatch_standing_task
from agentcore.standing_tasks.schedule import CronError, next_run_after, resolve_cron
from agentcore.standing_tasks.webhook import (
    enforce_webhook_rate_limit,
    extract_event_text,
    generate_webhook_secret,
    idempotency_lookup,
    idempotency_store,
    require_webhook_secret,
)

router = APIRouter(tags=["standing-tasks"])
hooks_router = APIRouter(prefix="/hooks", tags=["standing-hooks"])


def _require_cloud_folder(folder) -> None:
    if folder is None:
        raise NotFoundError("工作区不存在")
    if folder.local_root_id:
        raise ValidationError("站立任务仅支持云工作区（拒绝本地 folder）")


@router.post("/standing-tasks", response_model=StandingTaskSummary, status_code=201)
async def create_standing_task(
    body: CreateStandingTaskRequest,
    user: AuthUser,
    folders: FolderRepository = Depends(get_folder_repo),
    repo: StandingTaskRepository = Depends(get_standing_task_repo),
):
    folder = await folders.get_by_id(body.folder_id, user_id=user.user_id)
    _require_cloud_folder(folder)
    axes = (
        body.permission_axes.to_axes().to_dict()
        if body.permission_axes is not None
        else DEFAULT_PERMISSION_AXES.to_dict()
    )
    plaintext_secret: str | None = None
    if body.trigger_kind == "webhook":
        plaintext_secret, secret_hash = generate_webhook_secret()
        row = await repo.create(
            user_id=user.user_id,
            folder_id=body.folder_id,
            name=body.name,
            goal=body.goal,
            cron=None,
            permission_axes=axes,
            next_run_at=None,
            enabled=body.enabled,
            trigger_kind="webhook",
            webhook_id=new_id(),
            webhook_secret_hash=secret_hash,
        )
        return StandingTaskSummary.from_row(row, webhook_secret=plaintext_secret)

    try:
        cron = resolve_cron(cron=body.cron, preset=body.schedule_preset)
        next_at = next_run_after(cron, datetime.now(UTC))
    except CronError as e:
        raise ValidationError(str(e)) from e
    row = await repo.create(
        user_id=user.user_id,
        folder_id=body.folder_id,
        name=body.name,
        goal=body.goal,
        cron=cron,
        permission_axes=axes,
        next_run_at=next_at,
        enabled=body.enabled,
        trigger_kind="schedule",
    )
    return StandingTaskSummary.from_row(row)


@router.get("/standing-tasks", response_model=list[StandingTaskSummary])
async def list_standing_tasks(
    user: AuthUser,
    repo: StandingTaskRepository = Depends(get_standing_task_repo),
):
    rows = await repo.list_by_user(user.user_id)
    return [StandingTaskSummary.from_row(r) for r in rows]


@router.get("/standing-tasks/{task_id}", response_model=StandingTaskSummary)
async def get_standing_task(
    task_id: str,
    user: AuthUser,
    repo: StandingTaskRepository = Depends(get_standing_task_repo),
):
    row = await repo.get_by_id(task_id, user_id=user.user_id)
    if row is None:
        raise NotFoundError("站立任务不存在")
    return StandingTaskSummary.from_row(row)


@router.patch("/standing-tasks/{task_id}", response_model=StandingTaskSummary)
async def update_standing_task(
    task_id: str,
    body: UpdateStandingTaskRequest,
    user: AuthUser,
    repo: StandingTaskRepository = Depends(get_standing_task_repo),
):
    existing = await repo.get_by_id(task_id, user_id=user.user_id)
    if existing is None:
        raise NotFoundError("站立任务不存在")

    fields = body.model_fields_set
    kwargs: dict = {}
    if "name" in fields:
        kwargs["name"] = body.name
    if "goal" in fields:
        kwargs["goal"] = body.goal
    if "enabled" in fields:
        kwargs["enabled"] = body.enabled
    if "permission_axes" in fields and body.permission_axes is not None:
        kwargs["permission_axes"] = body.permission_axes.to_axes().to_dict()

    plaintext_secret: str | None = None
    target_kind = body.trigger_kind if "trigger_kind" in fields else existing.trigger_kind

    if "trigger_kind" in fields and body.trigger_kind != existing.trigger_kind:
        if body.trigger_kind == "webhook":
            # Switch schedule → webhook: clear cron clock, mint webhook identity.
            plaintext_secret, secret_hash = generate_webhook_secret()
            kwargs.update(
                {
                    "trigger_kind": "webhook",
                    "cron": None,
                    "next_run_at": None,
                    "webhook_id": new_id(),
                    "webhook_secret_hash": secret_hash,
                }
            )
        else:
            # Switch webhook → schedule: wipe webhook; require cron/preset.
            if "cron" not in fields and "schedule_preset" not in fields and "preset" not in fields:
                raise ValidationError("切换到定时触发时须提供 schedule_preset 或 cron")
            try:
                cron = resolve_cron(cron=body.cron, preset=body.schedule_preset)
                next_at = next_run_after(cron, datetime.now(UTC))
            except CronError as e:
                raise ValidationError(str(e)) from e
            kwargs.update(
                {
                    "trigger_kind": "schedule",
                    "cron": cron,
                    "next_run_at": next_at,
                    "webhook_id": None,
                    "webhook_secret_hash": None,
                }
            )
    elif target_kind == "schedule" and (
        "cron" in fields or "schedule_preset" in fields or "preset" in fields
    ):
        try:
            cron = resolve_cron(cron=body.cron, preset=body.schedule_preset)
            kwargs["cron"] = cron
            kwargs["next_run_at"] = next_run_after(cron, datetime.now(UTC))
        except CronError as e:
            raise ValidationError(str(e)) from e
    elif target_kind == "webhook" and (
        "cron" in fields or "schedule_preset" in fields or "preset" in fields
    ):
        raise ValidationError("webhook 任务不可设置 cron / schedule_preset")

    row = await repo.update(task_id, user_id=user.user_id, **kwargs)
    if row is None:
        raise NotFoundError("站立任务不存在")
    return StandingTaskSummary.from_row(row, webhook_secret=plaintext_secret)


@router.delete("/standing-tasks/{task_id}", response_model=StatusResponse)
async def delete_standing_task(
    task_id: str,
    user: AuthUser,
    repo: StandingTaskRepository = Depends(get_standing_task_repo),
):
    ok = await repo.delete(task_id, user_id=user.user_id)
    if not ok:
        raise NotFoundError("站立任务不存在")
    return StatusResponse()


@router.post(
    "/standing-tasks/{task_id}/run",
    response_model=TriggerStandingTaskResponse,
    status_code=202,
)
async def trigger_standing_task(
    task_id: str,
    user: AuthUser,
    repo: StandingTaskRepository = Depends(get_standing_task_repo),
    folders: FolderRepository = Depends(get_folder_repo),
):
    """立即跑一次（验收 / 收件箱重跑）。不推进 cron 时钟。"""
    task = await repo.get_by_id(task_id, user_id=user.user_id)
    if task is None:
        raise NotFoundError("站立任务不存在")
    folder = await folders.get_by_id(task.folder_id, user_id=user.user_id)
    _require_cloud_folder(folder)
    run_id = await dispatch_standing_task(
        task_id=task_id,
        user_id=user.user_id,
        advance_schedule=False,
        trigger_source="manual",
    )
    return TriggerStandingTaskResponse(run_id=run_id)


@router.post(
    "/standing-tasks/{task_id}/rotate-webhook-secret",
    response_model=RotateWebhookSecretResponse,
)
async def rotate_webhook_secret(
    task_id: str,
    user: AuthUser,
    repo: StandingTaskRepository = Depends(get_standing_task_repo),
):
    task = await repo.get_by_id(task_id, user_id=user.user_id)
    if task is None:
        raise NotFoundError("站立任务不存在")
    if task.trigger_kind != "webhook" or not task.webhook_id:
        raise ValidationError("仅 webhook 站立任务可轮换密钥")
    plaintext, secret_hash = generate_webhook_secret()
    row = await repo.update(
        task_id,
        user_id=user.user_id,
        webhook_secret_hash=secret_hash,
    )
    if row is None or not row.webhook_id:
        raise NotFoundError("站立任务不存在")
    return RotateWebhookSecretResponse(
        webhook_id=row.webhook_id,
        webhook_url=webhook_path(row.webhook_id),
        webhook_secret=plaintext,
    )


@router.get("/standing-task-runs", response_model=StandingTaskRunListResponse)
async def list_standing_task_runs(
    user: AuthUser,
    status: Literal["running", "succeeded", "failed", "awaiting_user"] | None = Query(None),
    unacked: bool = Query(False),
    limit: int = Query(50, ge=1, le=200),
    repo: StandingTaskRunRepository = Depends(get_standing_task_run_repo),
):
    items = await repo.list_for_user(
        user.user_id, status=status, limit=limit, unacked_only=unacked
    )
    badge = await repo.count_badge(user.user_id)
    return StandingTaskRunListResponse(
        items=[StandingTaskRunSummary.model_validate(r) for r in items],
        badge=badge,
    )


@router.post("/standing-task-runs/{run_id}/ack", response_model=StandingTaskRunSummary)
async def ack_standing_task_run(
    run_id: str,
    user: AuthUser,
    repo: StandingTaskRunRepository = Depends(get_standing_task_run_repo),
):
    row = await repo.ack(run_id, user_id=user.user_id)
    if row is None:
        raise NotFoundError("收件箱条目不存在")
    return StandingTaskRunSummary.model_validate(row)


@hooks_router.post(
    "/standing/{webhook_id}",
    response_model=TriggerStandingTaskResponse,
    status_code=202,
)
async def fire_standing_webhook(
    webhook_id: str,
    request: Request,
    repo: StandingTaskRepository = Depends(get_standing_task_repo),
    folders: FolderRepository = Depends(get_folder_repo),
    authorization: str | None = Header(None),
    x_agentcore_webhook_secret: str | None = Header(
        None, alias="X-AgentCore-Webhook-Secret"
    ),
    x_idempotency_key: str | None = Header(None, alias="X-Idempotency-Key"),
):
    """Public webhook fire — no user JWT; auth via shared secret header."""
    task = await repo.get_by_webhook_id(webhook_id)
    if task is None:
        raise NotFoundError("Webhook 不存在")
    require_webhook_secret(
        authorization=authorization,
        x_webhook_secret=x_agentcore_webhook_secret,
        expected_hash=task.webhook_secret_hash,
    )
    if not task.enabled:
        raise ValidationError("站立任务已停用")

    idem_key = (x_idempotency_key or "").strip() or None
    if idem_key:
        prior = idempotency_lookup(webhook_id, idem_key)
        if prior is not None:
            return TriggerStandingTaskResponse(run_id=prior)

    enforce_webhook_rate_limit(task.id)

    folder = await folders.get_by_id(task.folder_id, user_id=task.user_id)
    _require_cloud_folder(folder)

    body = await request.body()
    event_text = extract_event_text(body, content_type=request.headers.get("content-type"))

    run_id = await dispatch_standing_task(
        task_id=task.id,
        user_id=task.user_id,
        advance_schedule=False,
        event_text=event_text or None,
        trigger_source="webhook",
    )
    if idem_key:
        idempotency_store(webhook_id, idem_key, run_id)
    return TriggerStandingTaskResponse(run_id=run_id)
